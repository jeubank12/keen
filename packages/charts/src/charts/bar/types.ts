export type GroupMode = 'grouped' | 'stacked';

export type StackMode = 'normal' | 'percent';

export type Bar = {
  key: string;
  selector: (string | number)[];
  x: number;
  y: number;
  height: number;
  width: number;
  color: string;
};
