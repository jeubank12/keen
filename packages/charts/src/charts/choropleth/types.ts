export type Projection =
  | 'mercator'
  | 'orthographic'
  | 'azimuthalEqualArea'
  | 'equalEarth'
  | 'naturalEarth';

export type ProjectionState = {
  rotation: [number, number, number];
};

export type TooltipMeta = {
  color: string;
  label: string;
  value: number;
};
