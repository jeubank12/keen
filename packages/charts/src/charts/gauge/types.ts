export type InnerArc = {
  path: string;
  color: string;
  value: number;
};
