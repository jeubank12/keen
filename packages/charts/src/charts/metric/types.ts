export type MetricType = 'difference' | 'comparison' | 'simple';
