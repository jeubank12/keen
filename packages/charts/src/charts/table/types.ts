import { CeilType } from '@keen.io/ui-core';

export type HeaderCeilType = {
  key: string;
  value: CeilType;
};

export type FormatType = (value: CeilType) => CeilType;

export type FormatTypeObject = {
  [key: string]: FormatType;
};

export type ValueFormatter = FormatTypeObject | FormatType;
