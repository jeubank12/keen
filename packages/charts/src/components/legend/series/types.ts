export type DataSerie = {
  name: string;
  color: string;
};
