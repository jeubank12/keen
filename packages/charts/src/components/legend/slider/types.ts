export type Variant = 'vertical' | 'horizontal';

export type Mode = 'vertical' | 'horizontal';
