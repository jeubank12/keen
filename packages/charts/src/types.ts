import { Typography, TooltipMode } from '@keen.io/ui-core';
import { IconType } from '@keen.io/icons';

export type Dimension = {
  width: number;
  height: number;
};

export type Margins = {
  top: number;
  bottom: number;
  left: number;
  right: number;
};

export type Line = {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
};

export type Tick = {
  text?: string | number;
  size: number;
  x: number;
  y: number;
};

export enum Orientation {
  VERTICAL = 'vertical',
  HORIZONTAL = 'horizontal',
}

export type Labels = {
  enabled: boolean;
  typography: Typography;
  radiusAngle?: number;
};

export type Tooltip = {
  enabled: boolean;
  mode: TooltipMode;
};

export type Axis = {
  enabled: boolean;
  tickSize: number;
  tickPadding: number;
  stroke?: number;
  color?: string;
  labels?: Labels;
};

export type Grid = {
  enabled: boolean;
  color: string;
};

export type Icon = {
  color: string;
  type: IconType;
};

export type Metric = {
  label: {
    typography: Typography;
  };
  excerpt: {
    icons: {
      increase: Icon;
      decrease: Icon;
    };
    backgroundColor: string;
    typography: Typography;
  };
};

export type Theme = {
  colors: string[];
  metric?: Metric;
  tooltip?: Tooltip;
  axisX?: Axis;
  axisY?: Axis;
  gridX?: Grid;
  gridY?: Grid;
  labels?: Labels;
};

export type CommonChartSettings = {
  /** Theme using for chart styling */
  theme?: Theme;
  /** SVG height and width */
  svgDimensions?: Dimension;
  /** SVG margins */
  margins?: Margins;
};

export type TooltipState = {
  visible: boolean;
  selectors: { selector: DataSelector; color: string }[];
  x: number;
  y: number;
};

export type TimePrecision =
  | 'minute'
  | 'hour'
  | 'day'
  | 'week'
  | 'month'
  | 'year';

export type ScaleSettings = {
  type: 'time' | 'band' | 'linear';
  formatLabel?: (label: string | number) => string | number;
  precision?: TimePrecision;
};

export type DataSelector = (number | string)[];

export type LabelRotation = {
  anchor: string;
  radius: number;
  translateX: number;
  translateY: number;
};

export type GroupMode = 'grouped' | 'stacked';

export type StackMode = 'normal' | 'percent';
