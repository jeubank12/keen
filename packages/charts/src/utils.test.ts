import { scaleBand, scaleLinear, scaleUtc } from 'd3-scale';

import {
  getCenterPosition,
  getScaleValues,
  generateTicks,
  textFormat,
  getFromPath,
  calculateRange,
  EDGE_TICK_ALIGN,
} from './utils';

import { Orientation } from './types';

describe('@keen.io/charts - utils', () => {
  const domain = ['Sales', 'Marketing', 'E-commerce'];
  const RealDate = Date;

  beforeAll(() => {
    global.Date = jest.fn().mockImplementation(date => new RealDate(date));
    global.Date.UTC = jest
      .fn()
      .mockImplementation(date => new RealDate(date).getUTCDate());
  });

  afterAll(() => {
    global.Date = RealDate;
  });

  describe('calculateRange()', () => {
    const data = [
      { label: 'January', sale: -3, buy: 11, revenue: 30 },
      { label: 'February', sale: 12, buy: 3, revenue: 21 },
    ];

    it('should calculate minimum and maximum values for provided keys', () => {
      const { minimum, maximum } = calculateRange(data, 'auto', 'auto', [
        'sale',
        'revenue',
      ]);

      expect(minimum).toEqual(-3);
      expect(maximum).toEqual(30);
    });

    it('should return defined minimum and maximum values', () => {
      const minValue = -2;
      const maxValue = 10;

      const { minimum, maximum } = calculateRange([], minValue, maxValue, []);

      expect(minimum).toEqual(minValue);
      expect(maximum).toEqual(maxValue);
    });

    it('should return default value for minimum greater than 0', () => {
      const minValue = 3;
      const maxValue = 10;

      const { minimum } = calculateRange([], minValue, maxValue, []);

      expect(minimum).toEqual(0);
    });
  });

  describe('generateTicks()', () => {
    const rangeStart = 0;
    const rangeEnd = 0;
    const tickSize = 10;
    const firstDate = new Date('2020-01-01T00:00:00.000Z');
    const lastDate = new Date('2020-06-01T00:00:00.000Z');

    it('should create additional ticks', () => {
      const scale = scaleBand()
        .range([rangeStart, rangeEnd])
        .domain(domain);
      const ticks = generateTicks({
        x: 0,
        y: 0,
        scale,
        tickSize,
      });

      expect(ticks.length).toEqual(5);
    });

    it('should create ticks for vertical linear scale', () => {
      const scale = scaleLinear()
        .range([rangeStart, rangeEnd])
        .domain([0, 90]);

      const ticks = generateTicks({
        x: 0,
        y: 0,
        scale,
        tickSize,
        orientation: Orientation.VERTICAL,
      });

      expect(ticks).toMatchSnapshot();
    });

    it('should create ticks for horizontal linear scale', () => {
      const scale = scaleLinear()
        .range([rangeStart, rangeEnd])
        .domain([0, 90]);

      const ticks = generateTicks({
        x: 0,
        y: 0,
        scale,
        tickSize,
        orientation: Orientation.HORIZONTAL,
      });

      expect(ticks).toMatchSnapshot();
    });

    it('should create ticks for horizontal Utc scale', () => {
      const scale = scaleUtc()
        .range([rangeStart, rangeEnd])
        .domain([firstDate, lastDate]);

      const ticks = generateTicks({
        x: 0,
        y: 0,
        scale,
        tickSize,
        orientation: Orientation.HORIZONTAL,
      });

      expect(ticks).toMatchSnapshot();
    });

    it('should create ticks for vertical band scale', () => {
      const scale = scaleBand()
        .range([rangeStart, rangeEnd])
        .domain(domain);

      const ticks = generateTicks({
        x: 0,
        y: 0,
        scale,
        tickSize,
        orientation: Orientation.VERTICAL,
      });

      expect(ticks).toMatchSnapshot();
    });

    it('should create ticks for horizontal band scale', () => {
      const scale = scaleBand()
        .range([rangeStart, rangeEnd])
        .domain(domain);

      const ticks = generateTicks({
        x: 0,
        y: 0,
        scale,
        tickSize,
        orientation: Orientation.HORIZONTAL,
      });

      expect(ticks).toMatchSnapshot();
    });

    it('should increase size of edge ticks', () => {
      const scale = scaleBand()
        .range([rangeStart, rangeEnd])
        .domain(domain);
      const ticks = generateTicks({
        x: 0,
        y: 0,
        scale,
        tickSize,
      });

      const edgeTicks = ticks.slice(ticks.length - 2, ticks.length);
      edgeTicks.forEach(({ size }) => {
        expect(size).toEqual(tickSize + EDGE_TICK_ALIGN);
      });
    });
  });

  describe('getScaleValues()', () => {
    const firstDate = new Date('2020-01-01T00:00:00.000Z');
    const lastDate = new Date('2020-06-01T00:00:00.000Z');

    it('should return domain for band scale', () => {
      const scale = scaleBand().domain(domain);

      expect(getScaleValues(scale)).toEqual(domain);
    });

    it('should return ticks for linear scale', () => {
      const scale = scaleLinear()
        .range([0, 10])
        .domain([0, 10]);

      expect(getScaleValues(scale)).toMatchSnapshot();
    });

    it('should apply "timeModifier" for month precision', () => {
      const scale = scaleUtc()
        .range([0, 10])
        .domain([firstDate, lastDate]);

      expect(
        getScaleValues(scale, { type: 'time', precision: 'month' })
      ).toMatchSnapshot();
    });

    it('should apply "timeModifier" for minute precision', () => {
      const scale = scaleUtc()
        .range([0, 10])
        .domain([
          new Date('2020-01-06T15:00:00.000Z'),
          new Date('2020-01-06T15:10:00.000Z'),
        ]);

      expect(
        getScaleValues(scale, { type: 'time', precision: 'minute' })
      ).toMatchSnapshot();
    });

    it('should apply "timeModifier" for week precision', () => {
      const scale = scaleUtc()
        .range([0, 10])
        .domain([
          new Date('2020-01-01T15:00:00.000Z'),
          new Date('2020-01-30T15:00:00.000Z'),
        ]);

      expect(
        getScaleValues(scale, { type: 'time', precision: 'week' })
      ).toMatchSnapshot();
    });

    it('should apply "timeModifier" for year precision', () => {
      const scale = scaleUtc()
        .range([0, 10])
        .domain([
          new Date('2015-01-01T00:00:00.000Z'),
          new Date('2020-06-01T00:00:00.000Z'),
        ]);

      expect(
        getScaleValues(scale, { type: 'time', precision: 'year' })
      ).toMatchSnapshot();
    });

    it('should return ticks for UTC scale without applying time precision', () => {
      const scale = scaleUtc()
        .range([0, 10])
        .domain([firstDate, lastDate]);

      expect(getScaleValues(scale)).toMatchSnapshot();
    });
  });

  describe('getCenterPosition()', () => {
    it('should calculate center position for element in scale', () => {
      const scale = scaleBand()
        .range([0, 120])
        .domain(domain);

      const positionCalculator = getCenterPosition(scale);

      expect(positionCalculator('Sales')).toEqual(20);
      expect(positionCalculator('Marketing')).toEqual(60);
      expect(positionCalculator('E-commerce')).toEqual(100);
    });
  });

  describe('getFromPath()', () => {
    it('should extract value based on provided selector', () => {
      const person = {
        address: {
          city: 'New York',
        },
      };

      expect(getFromPath(person, ['address', 'city'])).toEqual('New York');
    });

    it('should extract value from collections based on provided selector', () => {
      const countries = [{ name: 'United States' }];

      expect(getFromPath(countries, [0, 'name'])).toEqual('United States');
    });
  });

  describe('textFormat()', () => {
    it('return value without formating', () => {
      const value = 50;
      const returnValue = textFormat(value);

      expect(returnValue).toEqual(50);
    });

    it('should convert "Date" instance to string format', () => {
      const value = new Date('2020-01-04');

      expect(textFormat(value)).toMatchInlineSnapshot(
        `2020-01-04T00:00:00.000Z`
      );
    });

    it('should apply "formatLabel" function', () => {
      const value = 'keen.io';
      const formatLabel = (label: string | number) => `@${label}`;
      const returnValue = textFormat(value, { type: 'band', formatLabel });

      expect(returnValue).toEqual('@keen.io');
    });
  });
});
