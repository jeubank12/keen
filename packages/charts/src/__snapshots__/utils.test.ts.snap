// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen.io/charts - utils generateTicks() should create ticks for horizontal Utc scale 1`] = `
Array [
  Object {
    "size": 10,
    "text": "Wed Jan 01 2020 01:00:00 GMT+0100 (Central European Standard Time)",
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": "Sat Feb 01 2020 01:00:00 GMT+0100 (Central European Standard Time)",
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": "Sun Mar 01 2020 01:00:00 GMT+0100 (Central European Standard Time)",
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": "Wed Apr 01 2020 02:00:00 GMT+0200 (Central European Summer Time)",
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": "Fri May 01 2020 02:00:00 GMT+0200 (Central European Summer Time)",
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": "Mon Jun 01 2020 02:00:00 GMT+0200 (Central European Summer Time)",
    "x": 0.5,
    "y": 0,
  },
]
`;

exports[`@keen.io/charts - utils generateTicks() should create ticks for horizontal band scale 1`] = `
Array [
  Object {
    "size": 10,
    "text": "Sales",
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": "Marketing",
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": "E-commerce",
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 14,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 14,
    "x": 0.5,
    "y": 0,
  },
]
`;

exports[`@keen.io/charts - utils generateTicks() should create ticks for horizontal linear scale 1`] = `
Array [
  Object {
    "size": 10,
    "text": 0,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 10,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 20,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 30,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 40,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 50,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 60,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 70,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 80,
    "x": 0.5,
    "y": 0,
  },
  Object {
    "size": 10,
    "text": 90,
    "x": 0.5,
    "y": 0,
  },
]
`;

exports[`@keen.io/charts - utils generateTicks() should create ticks for vertical band scale 1`] = `
Array [
  Object {
    "size": 10,
    "text": "Sales",
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": "Marketing",
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": "E-commerce",
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 14,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 14,
    "x": 0,
    "y": -0.5,
  },
]
`;

exports[`@keen.io/charts - utils generateTicks() should create ticks for vertical linear scale 1`] = `
Array [
  Object {
    "size": 10,
    "text": 0,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 10,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 20,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 30,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 40,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 50,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 60,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 70,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 80,
    "x": 0,
    "y": -0.5,
  },
  Object {
    "size": 10,
    "text": 90,
    "x": 0,
    "y": -0.5,
  },
]
`;

exports[`@keen.io/charts - utils getScaleValues() should return ticks for Utc scale 1`] = `
Array [
  2020-01-01T00:00:00.000Z,
  2020-02-01T00:00:00.000Z,
  2020-03-01T00:00:00.000Z,
  2020-04-01T00:00:00.000Z,
  2020-05-01T00:00:00.000Z,
  2020-06-01T00:00:00.000Z,
]
`;

exports[`@keen.io/charts - utils getScaleValues() should return ticks for linear scale 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
]
`;
